File: src/java/org/codehaus/testdox/intellij/TestClass.java
package org.codehaus.testdox.intellij;

import javax.swing.Icon;

import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiElement;

public class TestClass extends AbstractTestElement {

    private final String className;
    private final PsiClass psiClass;
    private final EditorApi editorApi;
    private final NameResolver nameResolver;

    public TestClass(String className, PsiClass psiClass, EditorApi editorApi, NameResolver resolver) {
        this.className = className;
        this.psiClass = psiClass;
        this.editorApi = editorApi;
        this.nameResolver = resolver;
    }

    public String displayString() {
        return new StringBuffer("<b>").append(nameResolver.getRealClassNameForDisplay(className)).append(":</b>").toString();
    }

    public PsiElement getPsiElement() {
        return psiClass;
    }

    public boolean jumpToPsiElement() {
        return editorApi.jumpToPsiClass(psiClass);
    }

    public Icon icon() {
        return (psiClass != null) ? IconHelper.getIcon(IconHelper.CLASS_ICON)
                                  : IconHelper.getLockedIcon(IconHelper.CLASS_ICON);
    }

    public void updatePresentation(Presentation presentation) {
        presentation.setEnabled(true);
    }

    public void rename(TestDoxController testDoxController) {
        editorApi.rename(getTestedClass());
    }

    public void delete(TestDoxController testDoxController) {
        editorApi.delete(getTestedClass());
    }

    public boolean isTestClass() {
        return psiClass != null && nameResolver.isTestClass(psiClass.getName());
    }

    public int compareTo(TestElement object) {
        if (object instanceof TestClass) {
            return displayString().compareTo(object.displayString());
        }
        return 1;
    }

    public boolean equals(Object object) {
        return ((object instanceof TestClass) && (compareTo((TestElement) object) == 0));
    }

    private PsiClass getTestedClass() {
        return editorApi.getPsiClass(nameResolver.getRealClassName(psiClass.getQualifiedName()));
    }
}

File: src/scala/org/codehaus/testdox/intellij/TestClass.scala
package org.codehaus.testdox.intellij

import com.intellij.openapi.actionSystem.Presentation
import com.intellij.psi.PsiClass

class TestClass(className: String, psiClass: PsiClass, editorApi: EditorApi, nameResolver: NameResolver)
    extends AbstractTestElement {

  def displayString: String = "<b>" + nameResolver.getRealClassNameForDisplay(className) + ":</b>"

  override val psiElement = psiClass

  override def jumpToPsiElement(): Boolean = editorApi.jumpToPsiClass(psiClass)

  def isTestClass = psiClass != null && nameResolver.isTestClass(psiClass.getName())

  val icon = {
    if (psiClass != null)
      Icons.getIcon(Icons.CLASS_ICON)
    else
      Icons.getLockedIcon(Icons.CLASS_ICON)
  }

  override def update(presentation: Presentation) {
    presentation.setEnabled(true);
  }

  override def rename(controller: TestDoxController) {
    editorApi.rename(testedClass);
  }

  override def delete(controller: TestDoxController) {
    editorApi.delete(testedClass);
  }

  override def compare(that: TestElement): Int = {
    if (that.isInstanceOf[TestClass])
      this.displayString.compareTo(that.displayString)
    else
      1
  }

  override def equals(o: Any) = o.isInstanceOf[TestClass] && compareTo(o.asInstanceOf[TestElement]) == 0

  private def testedClass: PsiClass = editorApi.getPsiClass(nameResolver.getRealClassName(psiClass.getQualifiedName()))
}
